# Lesson 1: Guided Practice  

Key learning results:
- I understand the how to run Python code from the command line, in the interpreter, and in an Jupyter notebook.
- I can declare a variable.
- I feel comfortable writing various expressions.
- I understand "Pythonic" code organization.
- I recall the syntax for basic control flow statements.

## Where are we?: A brief introduction to running Python code

Before we actually start writing code, we need to talk about what we do with the code we're writing. Broadly, there are two ways of running Python code: line-by-line (a.k.a. interactively) or from a file. 

### Running code from file
Running a file in Python is as simple as openning up a terminal/command line window, navigating to the folder containing the file, then entering the following command:

```
python3 <file_name>
```

If you only have one version of Python installed or if the default version on Python on your system is Python 3, then it is also sufficient to use this command:

```
python <file_name>
```

Using the first command, however, ensures that our code is being run in Python 3.X and not Python 2.X. While Python 3.X is the most up-to-date, Python 2.X remains popular, and the two versions aren't perfectly cross compatible. This issue likely won't come up until you start exploring different, more advanced packages available in Python.

If you feel uncomfortable navigating your computer from the terminal/command line, I suggest [working through this short tutorial.](https://blog.teamtreehouse.com/introduction-to-the-mac-os-x-command-line) You don't need to become an expert, but a basic understanding of how to interface with your computer through the command line.

Programs don't by default have a command line interface, and though building one out is fairly simple, it won't be covered in any of the Python Jams. Still, it's good to be aware of how to run programs from the command line.

### Running code interactively

Throughout these sessions, we'll be running code interactively either through IPython or through a Jupyter notebook. Both allow us to work through code line by line, evaluating our progress as we work. 

#### IPython

To access IPython, open up the terminal/command line and run this command:

```
ipython3
```

You should see the following text appear on screen:

```
Python 3.7.3 (default, Mar 27 2019, 16:54:48) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.5.0 -- An enhanced Interactive Python. Type '?' for help.
```

This text is followed by a prompt where you can type in Python code and see its results.

#### Your first program

Now that you're ready to write Python code, lets print out a simple Hello World statement. In Python, this requires a single line of code:

```
print("Hello, world!")
```

Type this into the IPython window we just opened, and you should see the text ```Hello, world!``` appear below where you just typed!

#### Jupyter notebooks

Another place where we can run code iteractively is a Jupyter notebook. Indeed, this document you're reading right now is a Jupyter notebook. Running code in a Jupyter notebook allows you to combine text (in Markdown), plots, and other media alongside your code. This ability can be helpful if you're using Python to produce an analysis you intend to send to others or if you would like to explain the code you're writing as you're writing it. Furthermore, Jupyter notebooks also allow us to export 

Below this text, there is an empty "code" cell where you can write and run Python code within this document. Try typing our Hello, World! statement in that cell and then run it by either pressing Shift + Return or pressing the run button in the navigation bar.

In [48]:
print('Hello, World!')

Hello, World!


The majority of Python Jams will be in Jupyter notebooks, however, you should also try running some code in IPython so that you feel comfortable with it. When working by yourself on a project, you'll likely find it easier to work in IPython than in Jupyter notebooks.

## Basics of Python Programming

The topics covered in this section may seem elementary, but are essential building blocks for writing or debugging more advanced programs. While the section will be focused on how these topics are implemented in Python, these elements will be available in any programming lanague, though possibly with a slightly different impelementation.

### Variables

Variables are symbolic links that allow you to keep track of information in your machine's memory while a program runs. When defining a variable, you give it a name that you will use later on to retrieve or update the information stored in that variable.

In Python, we define variables like this:

```
<variable name> = <value>
```

Let's try rewriting our Hello, World! statement above using a variable. First, define a variable named *message* with the value "Hello, World!" in the empty cell below then press Shift + Return. Notice that when you want to store a block of text (called a string) as a variable, you have to wrap the text block in quotation marks. In Python, you can use either single (') or double (") quotation marks to wrap strings.

In [49]:
message = "Hello, World!"

Now, we can use the print function from above to print out the value of our *message* variable with the command

```
print(message)
```

Try it in the empty cell below:

In [50]:
print(message)

Hello, World!


You can also check the value of a variable by simply typing its name into the interpreter. This is helpful when working on or debugging a program but will not produce any output if you're running a program from the command line. Check the value of our message variable below:

In [51]:
message

'Hello, World!'

### Data Types

If you've ever written code in another language, you may have felt like something was missing in our variable assignment statement above. Namely, in Python there is no need to declare the type of variable you're declaring before you declare it.  

What does this mean? In Java, for example, you would declare the variable above as"

```
String message = "Hello, World!"
```

That first keyword, "String", is used to let Java know that the data you're storing is a string of text. Python, however, doesn't require explicity stating what type of data you're storing in a variable when you declare it. This doesn't mean, however, that variables in Python don't have a specific type associated with them. Indeed, we can check the type of a variable by using the type function.

In [52]:
type(message)

str

"str" means that this variable is a string. In Python, there are four basic types of variables and number of subtypes. The most important ones to know are:  

- Numerics: Can encompass both integer and real numbers
    - Integers (exclusively for whole numbers)
    - Floats/Floating point decimals (for whole numbers or decimals)

In [53]:
x = 5
type(x)

int

In [54]:
y = 11.012017
type(y)

float

In [55]:
type(y)

float

- Strings: Represent text as a 'string' of characters (i.e. the comptuer sees 'Hello, World!' as an 'H' character followed by an 'e' character followed by an 'l' character, etc.). There are a number of subtypes of strings with specific applications that will be covered later.

In [56]:
text = 'Welcome to Python!'
type(text)

str

- Booleans: represent logical values, i.e. True or False

In [57]:
t_or_f = True
type(t_or_f)

bool

- "None": Indicates the absence of a value. In some cases, it's useful to be able to explicitly note that a value does not exist or was not provided by a user.

In [58]:
a = None
type(a)

NoneType

Now, try declaring a few variables for youself and checking their types using the ```type()``` function.

### Expressions

Now that you know how to declare variables and have seen some of the different types of variables, you're ready to start combining them to produce new values.  

#### Arithmetic

One extremely basic use for Python would be as a simple calculator. All the basic arithmetic expressions, addition (```+```), subtraction (```-```), multipliation (```*```), exponentiation (```**```)division (```/```) work basically as expected.

In [59]:
11 + 1

12

In [60]:
11 - 1

10

In [61]:
11 * 2

22

In [62]:
11 ** 2

121

In [63]:
12 / 2

6.0

In the last answer, notice that we divided one integer by another yet the output was a float (you can tell by the decimal point). Standard division in Python will always produce a float, regardless of the two values you're dividing. If for some reasons you'd rather have an integer result, you should use the integer division operator, ```//```.

In [64]:
12 // 2

6

In [65]:
11 // 2

5

In [66]:
-11 // 2

-6

In effect, integer division will divide the two values then round down to the nearest integer.  

Lastly, there is also a modulus operator, ```%```, that yields the remainder of an integer division operation.

In [67]:
3.3 % 3

0.2999999999999998

<div class="alert alert-block alert-warning">
<b>Whoa, isn't that answer wrong?</b><br />
In the above cell block, you may have noticed that Python returns a value of 0.2999999999999998 instead of 0.3. This issue occurs because of how computers must use an approximation to store decimals. In some scientific operations, these so-called "floating point errors" can cause an issue, and it's necesary to use the Python decimal type. If you're curious about this, I'd encourage you to <a href="https://docs.python.org/2/library/decimal.html">read the Python documentation on decimal types.</a>
</div>  



We can also chain operators using the standard mathematical order of operations.

In [68]:
10 * 2 + 2

22

Finally, we can also use variables in our expression either to store the result of our expressions or perform operations on our different varaibles.

In [69]:
a = 10
b = 2
a * b + b

22

In [70]:
c = a * b + b

In [71]:
print(c)

22


#### Strings

Python's built-in operators also provide an easy was to join two strings together using the addition operator.

In [72]:
'abc' + 'def'

'abcdef'

In [73]:
message1 = 'Hello '
message2 = 'world!'
message3 = message1 + message2
print(message3)

Hello world!


Notice, however, that attempting to add a string and a number will yield an error.

In [74]:
'one' + 1

TypeError: can only concatenate str (not "int") to str

#### Relational operators

This next set of operators are used to compare two values, and the output of an expression involving them is always a boolean. The include:

- ```>```: greater than
- ```>=```: greater that or equal to
- ``` <```: less than
- ```<=```: less than or equal to
- ```==```: equals
- ```!=```: not equals

All six of these can be used on numeric type data, and the final two (equals and not equals) can be used on string type data. Let's see them in action:

In [75]:
5 == 5

True

In [76]:
10 < 2

False

In [77]:
6 != 7

True

In [78]:
"Hello" == "Hello"

True

In [79]:
"Hello" == "Bonjour"

False

In [80]:
"hello" == "Hello"

False

Notice that string equality is case sensitive.

In the cell below, I've defined two variables, ```a``` and ```b```. Write three expressions in the empy cells to check wheter a is greater than zero, whether b is greater than or equal to zero, and whether a times b is less than zero.

In [81]:
a = 10
b = -1

In [82]:
#is a greater than zero? enter your code below this line

In [83]:
#is b greater than or equal to zero? enter your code below this line

In [84]:
#is a + b equal to zero? enter your code below this line

#### Logical operators

We can combine multiple expressions with relational operators using the logical operators ```and``` and ```or```. Consider this expression:

In [85]:
a = 20
b = 30

a > 25 and b > 25

False

This expression would only evaluate to True if both a and b were greater than 25.

In [86]:
a > 25 or b > 25

True

This expression will evaluate to True if either a or b was greater than 25. 

Finally, there is also a ```not``` operator that will negate a boolean value. For example:

In [87]:
a = 5
a > 0

True

In [88]:
not a > 0

False

The value of ```a``` is greater than zero, therefore the first expression evaluate to True. The not operator, however, negates this True value in the second expression, so the second expression produes False. Consider another example. Do you understand why the expression evaluates to be True?

In [89]:
a = 5
not a < 0

True

Similar to arithmetic, you can (and should!) use parentheses in your expressions to clarify order of operations. The next expression checks that a is either less than or equal to zero or greater than or equal to 5.

In [90]:
not (a > 5 or 0 < a)

False

The following expression will check whether a is greater than 0 or less than or equal to 5.

In [91]:
 not a > 5 or a < 0

True

## Pythonic Code Organization